ls: Implement --zero flag. (#2929)

pimzero opened the pull request:
Fixes #2929

(Full disclosure, I am very new to rust, and contributing mainly as a learning exercise. But I hope this change can be useful).

This flag can be used to provide a easy machine parseable output from
ls, as discussed in the GNU bug report
https://debbugs.gnu.org/cgi/bugreport.cgi?bug=49716.

There are some peculiarities with this flag:

 - Current implementation of GNU ls of the `--zero` flag implies some
   other flags. Those can be overridden by setting those flags after
   `--zero` in the command line.
 - This flag is not compatible with `--dired`. This patch is not 100%
   compliant with GNU ls: GNU ls `--zero` will fail if `--dired` and
   `-l` are set, while with this patch only `--dired` is needed for the
   command to fail.

We also add `--dired` flag to the parser, with no additional behaviour
change.

Testing done:
```
$ bash util/build-gnu.sh
 [...]
$ bash util/run-gnu-test.sh tests/ls/zero-option.sh
 [...]
 PASS: tests/ls/zero-option.sh
 ============================================================================
 Testsuite summary for GNU coreutils 9.1.36-8ec11
 ============================================================================
 # TOTAL: 1
 # PASS:  1
 # SKIP:  0
 # XFAIL: 0
 # FAIL:  0
 # XPASS: 0
 # ERROR: 0
 ============================================================================
```

sylvestre commented:
Cool for a first contrib.
Fails on Windows with:

```

---- test_ls::test_ls_zero stdout ----
current_directory_resolved: 
mkdir: C:\Users\RUNNER~1\AppData\Local\Temp\.tmpVlGog3\0-test-zero
touch: C:\Users\RUNNER~1\AppData\Local\Temp\.tmpVlGog3\1
test-zero
thread 'test_ls::test_ls_zero' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 123, kind: Uncategorized, message: "The filename, directory name, or volume label syntax is incorrect." }', tests\common\util.rs:656:40

```

pimzero commented:
Thanks, I suppose this issue is caused by Windows filesystem not supporting newline in filenames, I updated the test to only run with the `\n` in filenames in unix. Also I fixed the lint & spelling issues the CI found

sylvestre commented on tests/by-util/test_ls.rs:
```diff
@@ -796,6 +796,116 @@ fn test_ls_commas() {
+        }
+
+        scene
+            .ucmd()
```
please run rustfmt on this file :)

pimzero replied:
done in 79f93e39d1a300c77347be65f5075aa202bf07b0

sylvestre commented:
Well done:
`Warning: Congrats! The gnu test tests/ls/zero-option is no longer failing!`

tertsdiepraam commented on src/uu/ls/src/ls.rs:
```diff
@@ -750,6 +747,81 @@ impl Config {
+            options::quoting::LITERAL,
+        ];
+        let get_last =
+            |flag: &str| -> usize { options.indices_of(flag).and_then(|x| x.last()).unwrap_or(0) };
```
I think this can be simplified by simply using `index_of`. Because we don't allow the flag multiple times but we allow them to override themselves. This means that clap will completely disregard the earlier argument. See this playground link: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1e3cd4125a923537d4e65effccdd7d9d

tertsdiepraam replied:
Also I think you can assume that only ever one argument will be present from each of those lists (again due to overriding).

Although I think that assumption does not hold for `LONG_NO_GROUP/OWNER`.

pimzero replied:
Thanks, I now use this in 61f77e9b61812036b0b5804a2dee8d56cd28ef0c, also allowing multiple `--zero`, as GNU `ls` allows it.

tertsdiepraam commented on src/uu/ls/src/ls.rs:
```diff
@@ -750,6 +747,81 @@ impl Config {
+        //  - `--color=none`
+        //  - `--quoting-style=literal`
+        // Current GNU ls implementation allows `--zero` Behavior to be
+        // overridden by later flags.
```
Excellent comment 👍

pimzero replied:
Thanks !

tertsdiepraam reviewed (commented):
Cool! Nice work on all the edge cases.
